File: src/app/src/main/java/com/huawei/elibri/java/view/fragment/ProfileFragment.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2012-2020. All rights reserved.
 */

package com.huawei.elibri.java.view.fragment;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.RadioGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.google.android.material.textfield.TextInputEditText;
import com.huawei.agconnect.cloud.database.CloudDBZone;
import com.huawei.agconnect.cloud.database.CloudDBZoneObjectList;
import com.huawei.agconnect.cloud.database.CloudDBZoneSnapshot;
import com.huawei.agconnect.cloud.database.exceptions.AGConnectCloudDBException;
import com.huawei.elibri.R;
import com.huawei.elibri.java.SavedPreference;
import com.huawei.elibri.java.adapter.InterestAdapter;
import com.huawei.elibri.java.interfaces.DBOpenInterface;
import com.huawei.elibri.java.service.model.Profile;
import com.huawei.elibri.java.service.repository.CloudDbQueries;
import com.huawei.elibri.java.utility.DialogUtils;
import com.huawei.elibri.java.utility.Util;
import com.huawei.hmf.tasks.Task;

import java.util.HashMap;

/**
 * Displays user Profile (name, email, interest, etc)
 *
 * @author: nWX914751
 * @since: 02-11-2020
 */
public class ProfileFragment extends InterestBaseFragment implements View.OnClickListener {
    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_profile, container, false);
        Button save = view.findViewById(R.id.button_save);
        save.setOnClickListener(this);
        initializeUi(view);
        return view;
    }

    /**
     * Initializes UI
     *
     * @param view rootView
     */
    private void initializeUi(View view) {
        Util.setProgressBar(getContext());
        initializeDb(new DBOpenInterface() {
            @Override
            public void success(CloudDBZone cloudDBZone) {
                Util.stopProgressBar();
                mCloudDBZone = cloudDBZone;
                checkIfUserExists(view);
                queryMaxInterestId();
            }

            @Override
            public void failure() {
                Util.stopProgressBar();
            }
        });
    }

    /**
     * Checks if user data exists in DB or not
     *
     * @param rootView : rootView
     */
    private void checkIfUserExists(View rootView) {
        if (mCloudDBZone == null) {
            Log.w(TAG, "CloudDBZone is null, try re-open it");
            return;
        }
        SavedPreference sp = SavedPreference.getInstance(getContext());
        Task<CloudDBZoneSnapshot<Profile>> queryTask = CloudDbQueries.getInstance().getUserProfile(sp, mCloudDBZone);
        queryTask
                .addOnSuccessListener(
                        snapshot -> {
                            inflateUserData(rootView, snapshot);
                            Log.d(TAG, "Fetch user profile data");
                        })
                .addOnFailureListener(Throwable::getLocalizedMessage);
    }

    /**
     * Inflates user data in UI
     *
     * @param rootView view
     * @param snapshot query result
     */
    private void inflateUserData(View rootView, CloudDBZoneSnapshot<Profile> snapshot) {
        Profile profileResult = processProfileQueryResult(snapshot);
        if (profileResult != null) {
            RadioGroup gender = rootView.findViewById(R.id.rg_gender);
            TextInputEditText age = rootView.findViewById(R.id.tiet_age);
            age.setText(String.valueOf(profileResult.getAge()));
            if (profileResult.getGender()) {
                gender.check(R.id.radio_female);
            } else {
                gender.check(R.id.radio_male);
            }
            gender.setEnabled(false);
            getUserInterests(rootView);
            SavedPreference.getInstance(getContext().getApplicationContext())
                    .isSubscribed(profileResult.getSubscription());
        } else {
            queryInterestList(rootView, new HashMap<>());
        }
    }

    /**
     * Returns user Profile
     *
     * @param snapshot : Cloud DB result
     * @return user Profile
     */
    private Profile processProfileQueryResult(CloudDBZoneSnapshot<Profile> snapshot) {
        CloudDBZoneObjectList<Profile> bookInfoCursor = snapshot.getSnapshotObjects();
        Profile interest = null;
        try {
            while (bookInfoCursor.hasNext()) {
                interest = bookInfoCursor.next();
            }
        } catch (AGConnectCloudDBException e) {
            Log.w(TAG, e);
        } finally {
            snapshot.release();
        }
        return interest;
    }


    /**
     * Show interests in UI
     *
     * @param view : RootView
     */
    @Override
    protected void showInterestList(View view) {
        SavedPreference sp = SavedPreference.getInstance(getContext());
        RecyclerView mRecyclerView = view.findViewById(R.id.rv_interests);
        InterestAdapter mAdapter = new InterestAdapter(interests, true);
        LinearLayoutManager manager = new LinearLayoutManager(getActivity());
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(manager);
        mRecyclerView.setAdapter(mAdapter);
        TextView name = view.findViewById(R.id.tv_name);
        name.setText(sp.getName());
        TextView emailId = view.findViewById(R.id.tv_emailid);
        emailId.setText(sp.fetchEmailId());
    }


    /**
     * Inserts user Profile data
     *
     * @param rootView parent view
     */
    private void upsertBookInfos(View rootView) {
        RadioGroup gender = rootView.findViewById(R.id.rg_gender);
        SavedPreference sp = SavedPreference.getInstance(getContext());
        Profile profile = new Profile();
        TextInputEditText age = rootView.findViewById(R.id.tiet_age);
        if (!TextUtils.isEmpty(age.getText())) {
            profile.setAge(Byte.parseByte(age.getText().toString()));
        }
        profile.setName(sp.getName());
        profile.setGender(gender.getCheckedRadioButtonId() == R.id.radio_female);
        boolean subscription = sp.getIsSubscribed();
        profile.setSubscription(subscription);
        profile.setEmailid(sp.fetchEmailId());
        if (mCloudDBZone == null) {
            Log.w(TAG, "CloudDBZone is null, try re-open it");
            return;
        }
        Task<Integer> upsertTask = mCloudDBZone.executeUpsert(profile);
        upsertTask
                .addOnSuccessListener(cloudDBZoneResult -> {
                    upsertInterests();
                    Log.d(TAG, "Profile data added successfully");
                })
                .addOnFailureListener(
                        e -> {
                            DialogUtils.showPositiveAlertDialog(
                                    getActivity(), e.getLocalizedMessage(), () -> getActivity().finishAffinity());
                            e.getLocalizedMessage();
                        });
    }

    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.button_save) {
            upsertBookInfos(view.getRootView());
        }
    }

    /**
     * It will ask to user does he wants to exit the app
     */
    public void onResume() {
        super.onResume();
        showAlertExitDialog(getView());
    }
}

File: src/app/src/main/java/com/huawei/elibri/kotlin/view/fragment/ProfileFragment.kt
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2012-2020. All rights reserved.
 */
package com.huawei.elibri.kotlin.view.fragment

import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.RadioGroup
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.textfield.TextInputEditText
import com.huawei.agconnect.auth.AGConnectAuth
import com.huawei.agconnect.cloud.database.CloudDBZone
import com.huawei.agconnect.cloud.database.CloudDBZoneSnapshot
import com.huawei.agconnect.cloud.database.exceptions.AGConnectCloudDBException
import com.huawei.elibri.R
import com.huawei.elibri.kotlin.SavedPreference
import com.huawei.elibri.kotlin.adapter.InterestAdapter
import com.huawei.elibri.kotlin.interfaces.DBOpenInterface
import com.huawei.elibri.kotlin.interfaces.DialogClickInterface
import com.huawei.elibri.kotlin.service.model.Profile
import com.huawei.elibri.kotlin.service.repository.CloudDbQueries
import com.huawei.elibri.kotlin.utility.DialogUtils.showPositiveAlertDialog
import com.huawei.elibri.kotlin.utility.Util.setProgressBar
import com.huawei.elibri.kotlin.utility.Util.stopProgressBar
import com.huawei.hmf.tasks.Task


/**
 * Displays user Profile (name, email, interest, etc)
 *
 * @author: nWX914751
 * @since: 02-11-2020
 */
class ProfileFragment : InterestBaseFragment(), View.OnClickListener {
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_profile, container, false)
        val save = view.findViewById<Button>(R.id.button_save)
        save.setOnClickListener(this)
        initializeUi(view)
        return view
    }

    /**
     * Initializes UI
     *
     * @param view rootView
     */
    private fun initializeUi(view: View) {
        setProgressBar(context)
        initializeDb(object : DBOpenInterface {
            override fun success(cloudDBZone: CloudDBZone?) {
                stopProgressBar()
                mCloudDBZone = cloudDBZone
                Log.d("Cloiud Db : ","Cloud DBZone " + mCloudDBZone)
                queryMaxInterestId()
                checkIfUserExists(view)

            }

            override fun failure() {
                stopProgressBar()
            }
        })
    }

    /**
     * Checks if user data exists in DB or not
     *
     * @param rootView : rootView
     */
    private fun checkIfUserExists(rootView: View) {
        if (mCloudDBZone == null) {
            Log.w(TAG, "CloudDBZone is null, try re-open it")
            return
        }
        val sp = SavedPreference.getInstance(context!!)
        val queryTask: Task<CloudDBZoneSnapshot<Profile>> = CloudDbQueries.instance!!.getUserProfile(sp!!, mCloudDBZone!!)
        queryTask
            .addOnSuccessListener { snapshot: CloudDBZoneSnapshot<Profile> ->
                inflateUserData(rootView, snapshot)
                Log.d(TAG, "Fetch user profile data")
            }
            .addOnFailureListener {
                    obj: Exception -> obj.localizedMessage
            }
    }

    /**
     * Inflates user data in UI
     *
     * @param rootView view
     * @param snapshot query result
     */
    private fun inflateUserData(
        rootView: View,
        snapshot: CloudDBZoneSnapshot<Profile>
    ) {
        val profileResult =
            processProfileQueryResult(snapshot)
        if (profileResult != null) {
            val gender = rootView.findViewById<RadioGroup>(R.id.rg_gender)
            val age: TextInputEditText = rootView.findViewById(R.id.tiet_age)
            age.setText(profileResult.age.toString())
            if (profileResult.gender) {
                gender.check(R.id.radio_female)
            } else {
                gender.check(R.id.radio_male)
            }
            gender.isEnabled = false
            getUserInterests(rootView)
            SavedPreference.getInstance(context!!.applicationContext)!!.isSubscribed(profileResult.subscription)
        } else {
          //  queryInterestList(rootView, HashMap())
        }
    }

    /**
     * Returns user Profile
     *
     * @param snapshot : Cloud DB result
     * @return user Profile
     */
    private fun processProfileQueryResult(snapshot: CloudDBZoneSnapshot<Profile>): Profile? {
        val bookInfoCursor =
            snapshot.snapshotObjects
        var interest: Profile? = null
        try {
            while (bookInfoCursor.hasNext()) {
                interest = bookInfoCursor.next()
            }
        } catch (e: AGConnectCloudDBException) {
            Log.w(TAG, e)
        } finally {
            snapshot.release()
        }
        return interest
    }

    /**
     * Show interests in UI
     *
     * @param view : RootView
     */
    override fun showInterestList(view: View?) {
        val sp = SavedPreference.getInstance(context!!)
        val mRecyclerView: RecyclerView = view!!.findViewById(R.id.rv_interests)
        val mAdapter = InterestAdapter(interests, true)
        val manager = LinearLayoutManager(activity)
        mRecyclerView.setHasFixedSize(true)
        mRecyclerView.layoutManager = manager
        mRecyclerView.adapter = mAdapter
        val name = view.findViewById<TextView>(R.id.tv_name)
        name.text = sp!!.name
        val emailId = view.findViewById<TextView>(R.id.tv_emailid)
        emailId.text = sp.fetchEmailId()
    }

    /**
     * Inserts user Profile data
     *
     * @param rootView parent view
     */
    private fun upsertBookInfos(rootView: View) {
        val gender = rootView.findViewById<RadioGroup>(R.id.rg_gender)
        val sp = SavedPreference.getInstance(context!!)
        val profile = Profile()
        val age: TextInputEditText = rootView.findViewById(R.id.tiet_age)
        if (!TextUtils.isEmpty(age.text)) {
            profile.age = age.text.toString().toByte()
        }
        profile.name = sp!!.name
        profile.gender = gender.checkedRadioButtonId == R.id.radio_female
        val subscription: Boolean = sp.isSubscribed
        profile.subscription = subscription
        profile.emailid = sp!!.fetchEmailId()
        if (mCloudDBZone == null) {
            Log.w(TAG, "CloudDBZone is null, try re-open it")
            return
        }
        Log.d("ProfileFragment","Profile : "+profile.name+profile.age+profile.emailid+profile.gender+profile.subscription)
        if (AGConnectAuth.getInstance().currentUser!=null ){
            val upsertTask: Task<Int> = mCloudDBZone!!.executeUpsert(profile)
            upsertTask.addOnSuccessListener { cloudDBZoneResult ->
                upsertInterests()
            }.addOnFailureListener {e: java.lang.Exception ->
                Log.d("ProfileFragment","exception : "+e)
                showPositiveAlertDialog(
                    activity!!,
                    e.localizedMessage,
                    object : DialogClickInterface {
                        override fun positiveButton() {
                            activity!!.finishAffinity()
                        }
                    }
                )
        }

            }else{
            Log.d("Cloud DB","Not opened");
        }

    }

    override fun onClick(view: View) {
        if (view.id == R.id.button_save) {
            upsertBookInfos(view.rootView)
        }
    }

    /**
     * It will ask to user does he wants to exit the app
     */
    override fun onResume() {
        super.onResume()
        showAlertExitDialog(view!!)
    }
}